### PR TITLE
Adding automated metadata generation to the test-runner

### PR DESCRIPTION
**Pull request by neenjaw**
The evolving test-runner spec requires the addition of metadata keys to the results.json response to a test request to facilitate web UI/UX flows.  This adds a framework to be able to generate this automatically as long as test suites conform (detailed below).

At a high level this sets out to accomplish the revised interface re: exercism/automated-tests#57 (yet to be merged).

### CODE REVIEW

**Comment by neenjaw** [comment]
@angelikatyborska I realize this is probably a huge PR to ask you to look at (as I think I've touched almost every file in the repo), but what I am looking for is a sanity check particularly looking from the starting point of the [Elixir CLI for the escript](https://github.com/exercism/elixir-test-runner/pull/23/files#diff-2c7056e709d66debd32fd5e1d8c5d2640430db1e2a12cba7aaa575315a804b7ac) and from the [assertion ast parser](https://github.com/exercism/elixir-test-runner/pull/23/files#diff-318cfc259098dfa6ae4a4011b33c587961013bf3179e86a5966b67234375104a)

**Comment by neenjaw** [comment]
@iHiD I think I need your review because I changed the [dockerfile slightly](https://github.com/exercism/elixir-test-runner/pull/23/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557)

**Review by iHiD — APPROVED** [review]
Approving from the "Codeowner" POV. I've not read the code :)

**Comment by angelikatyborska** [comment]
I started reviewing but it will take a few hours in total to finish, and that might spread over a few days because I'm low in free time ATM.

**Comment by angelikatyborska** [comment]
I'm a bit confused. https://github.com/exercism/v3/issues/2535 - this is an issue that _proposes_ adding two new keys to the metadata, `cmd` and `expected`. I do not see anywhere that this was already accepted as a consensus, there aren't many comments, but there are some comments against. But we're already adding those keys to Elixir test runner? Am I misunderstanding something? @iHiD are we supposed to be doing that? Are those two new keys already blessed by you and will be supported in v3?

**Comment by iHiD** [comment]
>  @iHiD are we supposed to be doing that? Are those two new keys already blessed by you and will be supported in v3?

Yes they are.

For background/clarity, those keys previously could be added via a per-exercise `config.json` (this is how the test-runners worked for the research site). e.g. https://github.com/exercism/research_experiment_1/blob/master/exercises/ruby-1-a/.meta/config.json#L4

This proposal was to add them so that they can be output along with the results, meaning a config.json doesn't need to be written for each test. We have this working in Ruby, C# and Common-Lisp. @neenjaw was relatively against the idea (maybe too strong a phrase, but didn't really want to take it on) and wanted to explore other ideas. So I left the "making it official" until he'd had chance to attempt/approach it for Elixir. He's now done this and is happy with it, so there's now no remaining objections from test-runner-authors, so we'll make it official, and update the docs

(cc @ErikSchierboom - Can I leave you in charge of getting this done this week pls? I'm happy to help where needed)

**Comment by angelikatyborska** [comment]
> so there's now no remaining objections from test-runner-authors

I don't know about other languages, but in Elixir it seems to have consequences for all track maintainers, not just test-runner authors. The consequences being that each test is supposed to only have a single assertion, which I personally find a very controversial suggestion. I don't see at the moment how I can make that work for some of the exercises (mostly the new v3 concept exercises that use processes).

@iHiD Can this be optional? Or does every single exercise in v3 need to produce a `cmd` and `expected` in the test runner output?

**Comment by iHiD** [comment]
>  The consequences being that each test is supposed to only have a single assertion, which I personally find a very controversial suggestion. I don't see at the moment how I can make that work for some of the exercises (mostly the new v3 concept exercises that use processes).
> @iHiD Can this be optional? Or does every single exercise in v3 need to produce a cmd and expected in the test runner output?

Every exercise needs to have a cmd/expected, else the student won't be able to see why the tests fail. As they can't see the tests for Concept Exercises, this is their only exposure to what is happening.

However, commands can have multiple assertions. There's not a rule against that.  Tim suggested that we could handle the expected key as an array, which I'm happy to explore. Currently I believe tracks have separated different expects with line-breaks. In really complex cases, we can also just use the original format (of putting the cmd/expected in the exercise's `config.json` instead, rather than writing edge-case code to solve it).

I would, however, argue that for the vast majority of Exercism tests, having single assertions does make it easy for the student to have just one failing thing and to make it clear what that is, rather than having to read the tests to work out which bit passes and which bit fails. But I understand that there are definitely going to be exceptions to that.

**Comment by neenjaw** [comment]
So I did say that, but then backtracked.  I'll explain why I backtracked.

For some of our tests, we have a preamble of code before the assertion, then the assertions. 

```elixir
describe "function" do
  test "feature" do
    # preamble

    assert actual == expected
  end
end
```

For the command, it likely wouldn't make sense to just have the code results in the _actual_ value in the assertion, so the command became the `preamble + actual`, with the expected.  So then looking at our existing tests, we often use multiple assertions, which then would require some sort of array to keep track of them. Sure. Easy.

But how can this be reflected in the website? The command then maybe is multiple lines long, displayed once for each assertion? How can we show from the command/expected which one actually caused the test to fail?

> Now in elixir's test runner, when the test fails the message displayed is the message from ExUnit (which is quite nice to read), but it probably wouldn't help the user sort out which command was causing the failure without close inspection.

So this is why I then backtracked and did not include that in the updated spec proposed in exercism/automated-tests#57.  I don't think there is an easy way to determine which assertion has failed in a test block with multiple assertions to help guide the student to understand using the "command"/"expected" values. I think for the sake of clarity and the web UI/UX, one assertion just makes sense.  When the test fails, we know which assertion failed, we know what message to display, we know how to help the student interpret their failure. I don't think we have that product clarity with the option for multiple assertions in a test case.

File: exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    updated_results =
+      %{results | "tests" => updated}
+      |> Jason.encode!()
+
+    File.write!(result_json_file, updated_results)
+  end
+
+  defp add_metadata(%{"name" => name} = test, metadata) do
+    IO.inspect(binding(), label: "31")
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex:31]: LOL, I have a macro that inserts either `IO.inspect(binding(), label: <current line of code>)` or `|> IO.inspect(label: "31")` so they get littered in there for eventual removal :P

File: exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  defp add_test_output(%{"name" => name} = test, log) do
+    cond do
+      log[name] ->
+        output = log[name]
+
+        output =
+          cond do
+            String.length(output) > 500 -> String.slice(output, -3..-1) <> "..."
+            true -> output
+          end
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex:61]: It's what I get for not writing a unit test for it :P

[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex:61]: I think its a very old type

File: exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+      true ->
+        Map.put(test, "output", nil)
+    end
+  end
+
+  defp add_metadata_error_fields(test, reason \\ "metadata unavailable") do
+    test
+    |> Map.put("cmd", "unavailable")
+    |> Map.put("expected", "unavailable")
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex:73]: Probably, at this point they are "expected" so I wanted to put something.  null probably makes more sense in the long run once a spec is confirmed

File: exercism_test_helper/lib/exercism_test_helper/cli/command/test_transform.ex [hunk 4]
```diff
@@ -0,0 +1,19 @@
+defmodule ExercismTestHelper.CLI.Command.TestTransform do
+  def run(file, replace) do
+    output_file =
+      if replace do
+        file
+      else
+        path = Path.dirname(file)
+        name = Path.basename(file, ".exs")
+        Path.join(path, name <> "_transformed.exs")
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/exercism_test_helper/cli/command/test_transform.ex:9]: Old code, note sure, will look at it, thanks

**Comment by angelikatyborska** [comment]
@iHiD  writes:
> Every exercise needs to have a cmd/expected, else the student won't be able to see why the tests fail.

Won't we show them the `message` from `output.json`? We have to... (see below).

@neenjaw  writes:
> I don't think there is an easy way to determine which assertion has failed in a test block when there are multiple assertions. And even with multiple command and expected values, it wouldn't easily guide the student to the error using the cmd & expected values. 

I would like to argue that the nature of programming is such that there will be a lot of cases in which a test fails, but **even if it has a single assertion, that doesn't mean that the assertion failed**.

For example, the code in the preamble can raise exceptions. The only way for a student to debug such a test is to see the actual output from the test runner. Which means they **must** see the actual output to fix their tests. We cannot tell them "cmd: foo(), expected: "bar", got: nil" when their code is raising errors. That's not helpful.

Let's say I am totally new to Elixir and I used the wrong operator to try to concat strings. This code will compile, but will raise errors:

```elixir
defmodule TwoFer do
  def two_fer(name) do
    "One for " + name + ", one for my cat"
  end
end
```

Now, for such a submission, a test result would look like this:

```
{
  cmd: 'TwoFer.two_fer("Gilberto Barros")',
  expected: 'to be equal to "One for Gilberto Barros, one for me"',
  message: '  2) test a name given (TwoFerTest)\n     test/two_fer_test.exs:8\n ' +
    '    ** (ArithmeticError) bad argument in arithmetic expression\n   ' +
    '  code: assert(TwoFer.two_fer("Gilberto Barros") == "One for ' +
    'Gilberto Barros, one for me")\n     stacktrace:\n       ' +
    ':erlang.+("Gilberto Barros", "One for ")\n       (two_fer 0.1.0) ' +
    'lib/two_fer.ex:3: TwoFer.two_fer/1\n       ' +
    'test/two_fer_test.exs:10: (test)\n',
  name: 'test a name given',
  output: '',
  status: 'fail'
}
```

If the website only shows me this information, without the message, I will have absolutely no idea what I did wrong. Did I make a typo in the string? Did I use the wrong type of quote for a string? Am I supposed to write `return`?
```
cmd: 'TwoFer.two_fer("Gilberto Barros")',
expected: 'to be equal to "One for Gilberto Barros, one for me"',
status: 'fail'
```

I have to see the message to have a chance of figuring out what I did wrong:
```
  2) test a name given (TwoFerTest)
     test/two_fer_test.exs:8
     ** (ArithmeticError) bad argument in arithmetic expression
     code: assert(TwoFer.two_fer("Gilberto Barros") == "One for Gilberto Barros, one for me")
     stacktrace:
       :erlang.+("Gilberto Barros", "One for ")
       (two_fer 0.1.0) lib/two_fer.ex:3: TwoFer.two_fer/1
       test/two_fer_test.exs:10: (test)
```

**Comment by ErikSchierboom** [comment]
@iHiD It looks like @neenjaw has helped me out here: https://github.com/exercism/automated-tests/pull/57

File: .dockerignore [hunk 5]
```diff
@@ -24,6 +24,6 @@ erl_crash.dump
 *.ez
 
 # Ignore package tarball (built via "mix hex.build").
-exercism_formatter-*.tar
+exercism_test_helper-*.tar
 
 results.json
```
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply .dockerignore:29]: Exercism will happily ignore any additional files, but it does make sense to ignore them in Git too when trying to run things locally.

File: Dockerfile [hunk 6]
```diff
@@ -26,7 +26,7 @@ RUN mix test --no-compile
 
 # Build the escript
 RUN mix escript.build
-RUN mv exercism_formatter /opt/test-runner/bin
+RUN mv exercism_test_helper /opt/test-runner/bin
```
Comment (ErikSchierboom) [inline Dockerfile:29]: I don't know anything about Erlang, but are there different Docker images? As in, one that has just the runtime and one that also has the SDK to build? Ignore if this does not make any sense.

You could consider using a [Docker multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) to only include the files needed to run the rest runner.

**Review by ErikSchierboom — APPROVED** [review]
@neenjaw Are there end-to-end tests to verify that for a given exercise, the correct outpuf files are created? If not, I found having a couple of those tests was really helpful.

File: exercism_test_helper/mix.exs [hunk 7]
```diff
@@ -3,7 +3,7 @@ defmodule ExercismFormatter.MixProject do
 
   def project do
     [
-      app: :exercism_formatter,
+      app: :exercism_test_helper,
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/mix.exs:6]: Fixed, thanks!

File: .dockerignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (neenjaw) [reply .dockerignore:29]: files added to gitignore, thanks!

File: Dockerfile (hunk 6 above)
Comment (neenjaw) [reply Dockerfile:29]: escripts still require an erlang runtime on the system, so having multistage build doesn't win much..

Comment (neenjaw) [reply Dockerfile:29]: It's a good suggestion though in general to consider

Comment (ErikSchierboom) [reply Dockerfile:29]: Ah okay, just wanted to verify. Thanks for explaining!

**Comment by iHiD** [comment]
@angelikatyborska The student sees the `cmd` and the `message` if the test fails and the `cmd` and the `expected` if it passes. 

This is an example of this in practice from the research website:

Screenshot 2020-11-24 at 12 01 00

**Comment by angelikatyborska** [comment]
@iHiD Thanks for the screenshot, it helps a lot. But does that mean that `expected` is supposed to be the actual return value of `cmd`? This PR sets `expected` to strings that describe the assertion, not the return value of `cmd`. This is because assertions other than `assert x == y` would be impossible to correctly convey in such a format.

**Comment by neenjaw** [comment]
> This is because assertions other than assert x == y would be impossible to correctly convey in such a format.

With the changes I am making following our conversation, this may not be strictly true. I am working on a way to represent all the assertion functions as natural language.

File: exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/exercism_test_helper/cli/command/combine_json.ex:52]: fixed in revision

**Comment by neenjaw** [comment]
Ok, so this has been a large refactor.

Initally started to obtain the `cmd` and `expected` (_now defunct for the moment_) from the test via the AST (rather than text-parsing). This now obtains the test body via the AST and returns a properly formatted string from the AST.

Not all assertions are fully covered here yet, but they can be easily (I think) -- just a matter of entering the function clauses. 

There has also been an interface change to the escript CLI to update it and make it nicer to use (thanks for the nudge @angelikatyborska :tada:).  The `run.sh` script interface remains unchanged, but the internals updated.

**Review by iHiD — APPROVED** [review]

**Comment by neenjaw** [comment]
This *should* work. I've tested it in the container, and locally following the steps in the docker file.

To test with a solution of your choosing, you could use this script to clean up the files created by this test runner:

```bash
#!/usr/bin/bash

rm output output.json results.json error_log metadata.json ./test/*altered ./test/*bkp
rm -rf _build
```

File: exercism_test_helper/.tool-versions [hunk 8]
```diff
@@ -0,0 +1 @@
+elixir 1.10.4-otp-23
```
Comment (angelikatyborska) [inline exercism_test_helper/.tool-versions:1]: Erlang version should also be specified in this file. I was getting an Erlang crash when trying to build this project, and then I noticed my default Erlang is 22.
```suggestion
elixir 1.10.4-otp-23
erlang 23.1
```

On top of that, it's also necessary to have a `.tool-versions` file in the root directory of this project, with just the Erlang version (maybe it's enough to specify the Erlang version just in the root direcctory and it will also be used inside of the `exercism_test_helper` directory? not sure how asdf behaves when there are two `.tool-versions`)

If I build the script with Erlang 23 from the directory `exercism_test_helper` and move it to root, it can't run from the root directory when there's a different Erlang version there. I was getting this error:
```
$ ./bin/run.sh two-fer /Users/angelika/Exercism/elixir/two-fer .
parsing /Users/angelika/Exercism/elixir/two-fer/test/two_fer_test.exs for metadata
escript: exception error: undefined function exercism_test_helper_escript:main/1
  in function  escript:run/2 (escript.erl, line 758)
  in call from escript:start/1 (escript.erl, line 277)
  in call from init:start_em/1 
  in call from init:do_boot/3 
```

File: exercism_test_helper/lib/meta/style.ex [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    :flunk,
+    :refute,
+    :refute_in_delta,
+    :refute_receive,
+    :refute_received
+  ]
+
+  def format(ast), do: Macro.to_string(ast, &format/2)
+
+  # Custom formatting handler for `if` and `unless` expressions
```
Comment (angelikatyborska) [inline exercism_test_helper/lib/meta/style.ex:20]: Do `cond` and `case` also need special treatment here?

**Comment by angelikatyborska** [comment]
The fact that the test body will be extracted and presented to the student without the context of the whole file is going to affect how we write tests. For example: avoid set up blocks, avoid module attributes - all code and data needs to be in the test.

AFAIK most tests won't require any change, but there are some exceptions, e.g.:
- using a module attribute https://github.com/exercism/v3/blob/16894b8a6f2fb1be8610969e89389bfe3904c25f/languages/elixir/exercises/concept/rpn-calculator-output/test/rpn_calculator/output_test.exs#L32-L40 can be easily fixed
- defining a struct just for the tests https://github.com/exercism/v3/blob/6576904cbe3ddae26b23a418da8e012162980c4a/languages/elixir/exercises/concept/bread-and-potions/test/rpg_test.exs#L6-L20 cannot be fixed...

Since those restrictions depend heavily on the implementation of the test runner and the features of the testing library itself, my assumption is that they are track-specific and each track is supposed to deal with this problem on its own.

So my question is: how can we communicate those restrictions for test files among the whole Elixir track so that all active maintainers, current future ones, will be aware of them, even if they didn't get involved in maintaining the test runner? I think so far all guidelines about writing exercises are in docs that apply to the whole Exercism platform.

**Comment by angelikatyborska** [comment]
This one will hurt :see https://github.com/exercism/v3/blob/6576904cbe3ddae26b23a418da8e012162980c4a/languages/elixir/exercises/concept/basketball-website/test/basketball_website_test.exs#L4-L110

File: exercism_test_helper/lib/meta/assert_parser.ex [hunk 10]
```diff
@@ -0,0 +1,136 @@
+defmodule Meta.AssertParser do
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/meta/assert_parser.ex:1]: Ya, I need to stick it into a feature branch so that it doesn't get lost in the case at some point we want to use it or reflect on the code.

File: exercism_test_helper/lib/meta/test_parser/test.ex [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    # {preamble, assertions} = parse_test_block(test_block)
+    # {last_command, expected} = parse_assertion(assertions)
+    # command = combine_preamble(preamble, last_command)
+
+    %__MODULE__{
+      name: make_name(description, name),
+      # preamble: preamble,
+      # assertions: assertions,
+      # command: command,
+      # expected: expected,
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/meta/test_parser/test.ex:16]: Ya, see above, I just hadn't strict it into a feature branch yet

File: exercism_test_helper/scratch.ex [hunk 12]
```diff
@@ -0,0 +1,20 @@
+defmodule ScratchTest do
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/scratch.ex:1]: Ya, that one wasn't meant to be commited.

File: exercism_test_helper/lib/meta/assert_parser.ex (hunk 10 above)
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/meta/assert_parser.ex]: I've moved it into a branch

File: exercism_test_helper/lib/meta/test_parser/test.ex [hunk 13]
```diff
@@ -0,0 +1,87 @@
+defmodule Meta.TestParser.Test do
+  @derive {Jason.Encoder, only: [:name, :test_body]}
+  defstruct [:name, :preamble, :assertions, :command, :expected, :test_body]
+
+  def make(description, name, test_block) do
+    test_body = Meta.AssertParser.Term.determine(test_block) |> to_string()
```
[the remark this answers is not in the document]
Comment (neenjaw) [reply exercism_test_helper/lib/meta/test_parser/test.ex]: I renamed it to `CodeBlock`

File: exercism_test_helper/lib/meta/style.ex (hunk 9 above)
Comment (neenjaw) [reply exercism_test_helper/lib/meta/style.ex:20]: surprisingly no, iirc, but maybe I should add a test to verify

**Comment by neenjaw** [comment]
I added some tests for cond/case and made a few things more consistent (re: trailing `\n` and whatnot)

**Comment by neenjaw** [comment]
Re: your concerns noted above ([1](https://github.com/exercism/elixir-test-runner/pull/23#issuecomment-735214615), [2](https://github.com/exercism/elixir-test-runner/pull/23#issuecomment-735215895)):

I don't think we should worry about this at this time.  I think us showing the lead up code and whatnot is good enough for now.

**Comment by angelikatyborska** [comment]
Will there be a way for the student using the web editor to take a look at the whole test file in the browser? If yes, then I can agree with:

>  I think us showing the lead up code and whatnot is good enough for now. 

If no, then some of the test suites, like the basketball website, will be annoying (maybe even impossible?) to work with online because the lead up code in the test block doesn't show the input data.

**Comment by iHiD** [comment]
They will see test suites for Practice Exercises, but not Concept Exercises.
